### DISCUSSION TITLE
Idea: void datatype

### DISCUSSION THREAD

**Discussion by pie-flavor** [dotnet/csharplang#2767] [Language Ideas]
Right now, there is a gulf between methods that return something, and methods that do not return something. This seems very unlike the rest of C#'s unifiedness, in the same way that it eschewed Java's gulf between objects and primitives. I think it would be a good idea to define `void` as a type whose only valid value is `void`, whether this means defining a new zero-sized struct or just using `void` as an alias to `System.ValueTuple` with no parameters when used as a type. This would fulfill the same use as `Unit` in Kotlin, `()` in Rust, or `Void` in Java, and allow void methods to be passed where a `Func` or similar is expected.

**Comment by yaakov-h** [comment]
See #538, #1279, and #1604.

**Comment by ekolis** [comment]
I like this idea; it would allow generic types to take void as a type parameter, reducing duplication of code (e.g. `Action<int>` could be redefined as `Func<int, void>`).

**Comment by pie-flavor** [comment]
538 isn't what I'm talking about. `never` has no instances, `void` has one instance. The other two talk about some new `unit` type instead of just treating `void` as a type from now on.

**Comment by eelstork** [comment]
I would have uses for a feature in this direction. 
I already use an `@void` single-valued type in places; mainly because `void` functions do not partake expressions, which breaks flow.
Actual `void` functions need to be bridged to that, which is not ideal.

Type conversions from void would be handy.

Would be less enthused by an approach equating `void` to either an empty tuple or `null`. Let `void` be `void` and programmers decide the details of how they wish to handle it.
